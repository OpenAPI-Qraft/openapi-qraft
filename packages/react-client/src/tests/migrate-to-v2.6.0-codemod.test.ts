import type transformer from '../migrate-to-v2.6.0-codemod.js';
import path from 'node:path';
import { describe, expect, it } from 'vitest';

describe('migrate-to-v2.6.0-codemod', () => {
  const transformPath = path.resolve('./src/migrate-to-v2.6.0-codemod.ts');
  const genericsFixturePath =
    './src/tests/fixtures/migrate-to-v2.6.0-codemod/generics.tsx';
  const otherImportsFixturePath =
    './src/tests/fixtures/migrate-to-v2.6.0-codemod/other-imports.tsx';
  const mixedImportsFixturePath =
    './src/tests/fixtures/migrate-to-v2.6.0-codemod/mixed-imports.tsx';
  const customPackageFixturePath =
    './src/tests/fixtures/migrate-to-v2.6.0-codemod/custom-package.tsx';

  it('removes generics from API client', async () => {
    const fs = await import('node:fs');
    const jscodeshiftModule = await import('jscodeshift');
    const jscodeshift = jscodeshiftModule.withParser('tsx');
    const transformerFunction = await import(transformPath).then(
      (module) => module.default as typeof transformer
    );

    expect(
      transformerFunction(
        {
          source: fs.readFileSync(genericsFixturePath, 'utf-8'),
          path: genericsFixturePath,
        },
        // @ts-expect-error - wrong types
        {
          jscodeshift,
          j: jscodeshift,
          stats: () => {},
        },
        {}
      )
    ).toMatchSnapshot('generics fixture');
  });

  it('does not modify files where qraftAPIClient is imported from another package', async () => {
    const fs = await import('node:fs');
    const jscodeshiftModule = await import('jscodeshift');
    const jscodeshift = jscodeshiftModule.withParser('tsx');
    const transformerFunction = await import(transformPath).then(
      (module) => module.default as typeof transformer
    );

    const source = fs.readFileSync(otherImportsFixturePath, 'utf-8');
    const result = transformerFunction(
      {
        source,
        path: otherImportsFixturePath,
      },
      // @ts-expect-error - wrong types
      {
        jscodeshift,
        j: jscodeshift,
        stats: () => {},
      },
      {}
    );

    // Source code should not be modified
    expect(result).toEqual(source);
    expect(result).toMatchSnapshot('other imports fixture');
  });

  it('correctly handles mixed imports', async () => {
    const fs = await import('node:fs');
    const jscodeshiftModule = await import('jscodeshift');
    const jscodeshift = jscodeshiftModule.withParser('tsx');
    const transformerFunction = await import(transformPath).then(
      (module) => module.default as typeof transformer
    );

    expect(
      transformerFunction(
        {
          source: fs.readFileSync(mixedImportsFixturePath, 'utf-8'),
          path: mixedImportsFixturePath,
        },
        // @ts-expect-error - wrong types
        {
          jscodeshift,
          j: jscodeshift,
          stats: () => {},
        },
        {}
      )
    ).toMatchSnapshot('mixed imports fixture');
  });

  it('supports custom package name option', async () => {
    const fs = await import('node:fs');
    const jscodeshiftModule = await import('jscodeshift');
    const jscodeshift = jscodeshiftModule.withParser('tsx');
    const transformerFunction = await import(transformPath).then(
      (module) => module.default as typeof transformer
    );

    // Check that code is not modified with default settings (not our package)
    const source = fs.readFileSync(customPackageFixturePath, 'utf-8');
    const defaultResult = transformerFunction(
      {
        source,
        path: customPackageFixturePath,
      },
      // @ts-expect-error - wrong types
      {
        jscodeshift,
        j: jscodeshift,
        stats: () => {},
      },
      {}
    );

    // Code should not be modified with default settings
    expect(defaultResult).toEqual(source);

    // Check that code is modified when correct package name is specified
    const customResult = transformerFunction(
      {
        source,
        path: customPackageFixturePath,
      },
      // @ts-expect-error - wrong types
      {
        jscodeshift,
        j: jscodeshift,
        stats: () => {},
      },
      {
        packageName: '@custom/package',
      }
    );

    // Code should be modified when correct package name is specified
    expect(customResult).not.toEqual(source);
    expect(customResult).toMatchSnapshot('custom package fixture');
  });

  it(
    'compatible jscodeshift runner',
    async () => {
      // @ts-expect-error - no types
      const jscodeshift = await import('jscodeshift/src/Runner').then(
        ({ run }) => run
      );

      expect(
        await jscodeshift(transformPath, [genericsFixturePath], {
          dry: true,
          print: false,
          verbose: 0,
          babel: true,
          parser: 'tsx',
        })
      ).toMatchObject({
        error: 0,
        nochange: 0,
        ok: 1,
        skip: 0,
        stats: {},
      });

      // Check with custom package name
      expect(
        await jscodeshift(transformPath, [customPackageFixturePath], {
          dry: true,
          print: false,
          verbose: 0,
          babel: true,
          parser: 'tsx',
          packageName: '@custom/package',
        })
      ).toMatchObject({
        error: 0,
        nochange: 0,
        ok: 1,
        skip: 0,
        stats: {},
      });
    },
    { timeout: 10_000 }
  );
});
