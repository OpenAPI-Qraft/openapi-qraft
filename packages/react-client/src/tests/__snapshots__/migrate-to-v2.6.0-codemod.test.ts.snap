// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate-to-v2.6.0-codemod > correctly handles mixed imports > mixed imports fixture 1`] = `
"// Import from our package
import { qraftAPIClient } from '@openapi-qraft/react-client';
// Import another function with the same name from another package
import { qraftAPIClient as otherClient } from 'other-library';

// This call should be modified (it uses our import)
const api = qraftAPIClient(services, callbacks, options);

// This call should NOT be modified (it uses another import)
const otherApi = otherClient<OtherServices, OtherCallbacks>(
  otherServices,
  otherCallbacks,
  otherOptions
);

interface Services {
  service1: {
    operation1: {
      schema: any;
    };
  };
}

interface Callbacks {
  operation1: () => void;
}

interface OtherServices {
  otherService: {
    otherOperation: {
      schema: any;
    };
  };
}

interface OtherCallbacks {
  otherOperation: () => void;
}
"
`;

exports[`migrate-to-v2.6.0-codemod > does not modify files where qraftAPIClient is imported from another package > other imports fixture 1`] = `
"// Import from another package, should not be modified
import { qraftAPIClient } from 'other-library';

// Same identifier but different source
const api = qraftAPIClient<Services, Callbacks>(services, callbacks, options);

// Local function with the same name
function qraftAPIClient<T, K>(a: T, b: K, c: any) {
  return { a, b, c };
}

// Local function call should not be modified
const localApi = qraftAPIClient<number, string>(1, "test", {});

interface Services {
  service1: {
    operation1: {
      schema: any;
    };
  };
}

interface Callbacks {
  operation1: () => void;
} "
`;

exports[`migrate-to-v2.6.0-codemod > removes generics from API client > generics fixture 1`] = `
"import { qraftAPIClient } from '@openapi-qraft/react-client';

const services: Services = {
  service1: {
    operation1: {
      schema: {},
    },
  },
};

const callbacks: Callbacks = {
  operation1: () => {},
};

const options = {
  requestFn: () => Promise.resolve({}),
  baseUrl: 'https://api.example.com',
  queryClient: {},
};

const requestFn = () => Promise.resolve({});
const baseUrl = 'https://api.example.com';
const queryClient = {};

const api = qraftAPIClient(services, callbacks, options);

const api2 = qraftAPIClient(services, callbacks, {
  requestFn,
  baseUrl,
  queryClient,
});

const api3 = qraftAPIClient(services, callbacks, {
  queryClient,
});

interface Services {
  service1: {
    operation1: {
      schema: any;
    };
  };
}

interface Callbacks {
  operation1: () => void;
}
"
`;

exports[`migrate-to-v2.6.0-codemod > supports custom package name option > custom package fixture 1`] = `
"// Import from custom package (not standard '@openapi-qraft/react-client')
import { qraftAPIClient } from '@custom/package';

// This call should be modified when correct package name is specified in options
const api = qraftAPIClient(services, callbacks, options);

// Additional call
const api2 = qraftAPIClient(services, callbacks, {
  requestFn,
  baseUrl,
  queryClient,
});

interface Services {
  service1: {
    operation1: {
      schema: any;
    };
  };
}

interface Callbacks {
  operation1: () => void;
}
"
`;
